pkg/controller/deployment/util/deployment_util.go

```diff
@@ -714,6 +715,17 @@ func GetAvailableReplicaCountForReplicaSets(replicaSets []*apps.ReplicaSet) int3
 	return totalAvailableReplicas
 }
 
+// GetTerminatingReplicaCountForReplicaSets returns the number of terminating pods for all replica sets.
+func GetTerminatingReplicaCountForReplicaSets(replicaSets []*apps.ReplicaSet) int32 {
+	terminatingReplicas := int32(0)
+	for _, rs := range replicaSets {
+		if rs != nil {
+			terminatingReplicas += ptr.Deref(rs.Status.TerminatingReplicas, 0)
```

Comment:
Defaulting a nil here to 0 unconditionally is not correct.

If this replicaset has been synced by the replicaset controller at all (status.observedGeneration != 0), and TerminatingReplicas is nil, we cannot reason about the sum of terminatingReplicas across the replicasets and have to return nil from GetTerminatingReplicaCountForReplicaSets

```suggestion
			if rs.Status.ObservedGeneration != 0 && rs.Status.TerminatingReplicas == nil {
				// we cannot sum TerminatingReplicas from replicasets that have been synced by the replicaset controller but do not have TerminatingReplicas set
				return nil
			}
			// sum TerminatingReplicas reported by the replicaset controller (or 0 for replicasets not yet synced by the controller)
			terminatingReplicas += ptr.Deref(rs.Status.TerminatingReplicas, 0)
```

Reply:
This seems reasonable at first glance. I will try to update the code handling for this in both PRs and test it.

Reply:
I have updated both PRs to handle the pointer fields. It seems to work fine when we check the ObservedGeneration as we will get non nil fields during normal operations. 

If we are unable to asses the number of terminating pods, the deployment will stop progressing until we are able to make that decision again. As mentioned in https://github.com/kubernetes/kubernetes/pull/128546#discussion_r1833333920, this should almost never happen.

Reply:
I *think* the current implementation does what we want functionally, but is sort of complicated to read because it's trying to accumulate the list of replicaset names we don't have terminating counts for.

I think this can be much simpler like the following:
```go
terminatingReplicas := int32(0)
for _, rs := range replicaSets {
  switch {
    case rs == nil:
      // No-op

    case rs.Status.ObservedGeneration == 0 && rs.Status.TerminatingReplicas == nil:
  		// Replicasets that have never been synced by the controller don't contribute to TerminatingReplicas

    case rs.Status.TerminatingReplicas == nil:
      // If any replicaset synced by the controller hasn't reported TerminatingReplicas, we cannot calculate a sum
      return nil

    default:
      terminatingReplicas += *rs.Status.TerminatingReplicas
  }
}
return &terminatingReplicas
```

Reply:
I think this is the only outstanding comment

Reply:
Updated to use the above snippet for better clarity. Thx!